packages/portfolio-deploy/src/contract-control.js

```diff
@@ -0,0 +1,203 @@
+/**
+ * @file contract-control: install, upgrade, terminate, reset chain-storage,
+ * for an orchestration contract in a branch of agoricNames / chain-storage.
+ */
+
+import { E } from '@endo/far';
+import { makeTracer } from '@agoric/internal/src/debug.js';
+import { Fail, q } from '@endo/errors';
+import { M } from '@endo/patterns';
+
+const trace = makeTracer('CCtrl');
+
+/**
+ * @import {OrchestrationPowers} from '@agoric/orchestration';
+ * @import {Zone} from '@agoric/base-zone';
+ * @import {ContractStartFunction} from '@agoric/zoe/src/zoeService/utils';
+ * @import {IssuerKeywordRecord} from '@agoric/zoe';
+ * @import {Remote} from '@agoric/internal';
+ * @import {Board, NameHubKit} from '@agoric/vats';
+ * @import {CopyRecord} from '@endo/pass-style';
+ */
+
+const IssuerShape = M.remotable('Issuer');
+const InstallationShape = M.remotable('Installation');
+const KitShape = M.record();
+const BundleIdShape = M.string();
+const VatUpgradeResultsShape = M.splitRecord({ incarnationNumber: M.number() });
+
+const iface = M.interface('ContractControl', {
+  install: M.callWhen(M.string()).returns({
+    boardId: M.string(),
+    installation: InstallationShape,
+  }),
+  start: M.callWhen(
+    M.splitRecord(
+      { installation: InstallationShape },
+      {
+        issuers: M.recordOf(M.string(), IssuerShape),
+        dataPrivateArgs: M.record(),
+      },
+    ),
+  ).returns(KitShape),
+  deploy: M.callWhen(
+    M.splitRecord(
+      { bundleId: BundleIdShape },
+      {
+        issuers: M.recordOf(M.string(), IssuerShape),
+        dataPrivateArgs: M.record(),
+      },
+    ),
+  ).returns(KitShape),
+  upgrade: M.callWhen(BundleIdShape).returns(VatUpgradeResultsShape),
+  terminate: M.callWhen(M.string()).returns(),
+  pruneChainStorage: M.callWhen(
+    M.recordOf(M.string(), M.arrayOf(M.string())),
+  ).returns(M.number()),
+});
+
+/**
+ * @template {ContractStartFunction} SF
+ * @typedef {StartedInstanceKit<SF> & {
+ *   label: string;
+ *   privateArgs: Parameters<SF>[1];
+ * }} UpgradeKit
+ */
+
+/**
+ * @param {Zone} zone
+ */
+export const prepareContractControl = zone => {
+  return zone.exoClass(
+    'ContractControl',
+    iface,
+    // TODO: promises can't be stored durably
+    /**
+     * @template {ContractStartFunction} SF
+     * @param {{
+     *   name: string,
+     *   agoricNamesAdmin: ERef<NameHubKit['nameAdmin']>,
+     *   board: ERef<Board>,
+     *   zoe: ERef<ZoeService>,
+     *   startUpgradable: ERef<StartUpgradable>,
+     *   orchestrationPowers: OrchestrationPowers & {
+     *     storageNode: Remote<StorageNode>;
+     *   },
+     *   kit?: (StartedInstanceKit<SF> & { privateArgs: Parameters<SF>[1] }),
```

Comment:
Suggestion: Rename field `kit` to something less generic.
```suggestion
     *   instanceKit?: (StartedInstanceKit<SF> & { privateArgs: Parameters<SF>[1] }),
```